docs/flip/0001-default-dtype.md

```diff
@@ -0,0 +1,102 @@
+# FLIP: Default dtypes
+
+
+- Start Date: 2022-01-11
+- FLIP PR: [#1776](https://github.com/google/flax/pull/1776)
+- FLIP Issue: [#1777](https://github.com/google/flax/issues/1777)
+
+
+## Summary
+
+This FLIP proposes to replace the default dtype which is currently hard-coded to float32, and instead use the JAX type promotion results to derive a default dtype from the input and parameters of a layer.
+
+
+## Motivation
+
+Currently, linen Modules always produce float32 outputs regardless of input and parameter dtypes. Half-precision types like float16 and bfloat16 are supported by explicitly passing the half-precision type to each layer. The way this is currently implemented is that each layer has a dtype argument with float32 as the default value. The layer guarantees that this dtype will be the return type of the result returned by `__call__`.
+
+The current behavior is problematic and results in silent bugs especially for dtypes that do not fit inside float32 (complex, float64). Also the linen dtype behavior is significantly different from how NumPy and by extension JAX handle dtypes.
+
+
+### Dtypes in JAX
+
+JAX uses a NumPy inspired [dtype promotion](https://github.com/google/jax/blob/main/jax/_src/dtypes.py) mechanism as explained [here](https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=lattice#type-promotion-semantics). The type promotion rules are summarized by the following type lattice:
+
+![JAX type promotion lattice](https://jax.readthedocs.io/en/latest/_images/type_lattice.svg)
+
+
+
+## Dtypes in Linen
+
+Beside input arguments, state and in particular parameters could affect dtype promotion. For example: We might feed a float64 input to a Dense layer with float32 parameters. Currently the result would be truncated to float32. If the input is a complex number the result is even worse because the imaginary part will be silently dropped when casting to float32.
+
+By using the dtype promotion rules already available in JAX we can avoid this issue. A public api is available called `jax.numpy.result_dtype(*args)` which returns the dtype that JAX would promote the given arguments in accordance with the type promotion lattice. For linen layers the arguments would be the layer inputs together with the parameters so for example this would be inputs, kernel, and bias for a linear layer.
+
+
+# Implementation
+
+A simplified example implementation:
+
+
+```python
+def promote_arrays(*xs, dtype):
+ if dtype is None:
+   dtype = jnp.result_type(*jax.tree_leaves(xs))
+ return jax.tree_map(lambda x: jnp.asarray(x, dtype), xs)
```

Comment:
A problem with this implementation is that if I now have a layer such as `Dense(dtype=float32, param_dtype=float64)` and I feed this input some real and complex data `inputr=random(..., dtype=float64), inputc=random(..., complex128)`, it is impossible to express _perform the computation in single precision, be it real or complex_.

I think that in general it is never ok to drop a imaginary part silently. 
Unfortunately, there is no standard API to express `maybe_complex_with_precision`.

Reply:
> I think that in general it is never ok to drop a imaginary part silently.

I'm inclined to agree but I think this is really a problem that should be fixed in NumPy/JAX. I think the JAX devs would at least consider making this a warning/error but NumPy does not want to be backwards incompatible unfortunately. After all it doesn't help much if Flax has custom logic to avoid this but user could can still drop it silently.

What you could do in this case is cast down the params to single precision before you use the layer. Either with a nn.map_variables or before calling model.apply. This is actually pretty common when doing inference (typically float32 -> (b)float16, but could also be float64 -> float32) and comes with an additional speed boost because the model can load the lower precision weights for each inference call and only has to do the downcast once.

Reply:
Yeah, I agree with you that this must be fixed at the numpy level.
So, to the extent of what is possible without changing numpy, the design you propose looks good to me.